packages/core/src/domain/session/sessionStoreManager.ts

```diff
@@ -0,0 +1,256 @@
+   * - if the session is not active, clear the session store and expire the session cache
+   */
+  function watchSession() {
+    processSessionStoreOperations(
+      {
+        process: (sessionState) => (!isActiveSession(sessionState) ? {} : undefined),
+        after: synchronizeSession,
+      },
+      sessionStore
+    )
+  }
+
+  function synchronizeSession(sessionState: SessionState) {
+    if (!isActiveSession(sessionState)) {
+      sessionState = {}
+    }
+    if (hasSessionInCache()) {
+      if (isSessionInCacheOutdated(sessionState)) {
+        expireSessionInCache()
+      } else {
+        sessionCache = sessionState
+      }
+    }
+    return sessionState
+  }
+
+  function expandOrRenewCookie(sessionState: SessionState) {
+    const { trackingType, isTracked } = computeSessionState(sessionState[productKey])
+    sessionState[productKey] = trackingType
+    if (isTracked && !sessionState.id) {
+      sessionState.id = generateUUID()
+      sessionState.created = String(dateNow())
+    }
+    return isTracked
+  }
+
+  function hasSessionInCache() {
+    return sessionCache[productKey] !== undefined
+  }
+
+  function isSessionInCacheOutdated(sessionState: SessionState) {
+    return sessionCache.id !== sessionState.id || sessionCache[productKey] !== sessionState[productKey]
+  }
+
+  function expireSessionInCache() {
+    sessionCache = {}
+    expireObservable.notify()
+  }
+
+  function renewSessionInCache(sessionState: SessionState) {
+    sessionCache = sessionState
+    renewObservable.notify()
+  }
+
+  function retrieveActiveSession(): SessionState {
+    const session = retrieveSession()
+    if (isActiveSession(session)) {
+      return session
+    }
+    return {}
+  }
+
+  function isActiveSession(sessionDate: SessionState) {
+    // created and expire can be undefined for versions which was not storing them
+    // these checks could be removed when older versions will not be available/live anymore
+    return (
+      (sessionDate.created === undefined || dateNow() - Number(sessionDate.created) < SESSION_TIME_OUT_DELAY) &&
+      (sessionDate.expire === undefined || dateNow() < Number(sessionDate.expire))
+    )
+  }
+
+  return {
+    expandOrRenewSession: throttle(expandOrRenewSession, storeAccessOptions.pollDelay).throttled,
+    expandSession,
+    getSession: () => sessionCache,
+    renewObservable,
+    expireObservable,
+    expire: () => {
+      clearSession()
+      synchronizeSession({})
+    },
+    stop: () => {
+      clearInterval(watchSessionTimeoutId)
+    },
+  }
+}
+
+type Operations = {
+  process: (sessionState: SessionState) => SessionState | undefined
+  after?: (sessionState: SessionState) => void
+}
+
+const bufferedOperations: Operations[] = []
+let ongoingOperations: Operations | undefined
+
+export function processSessionStoreOperations(operations: Operations, sessionStore: SessionStore, numberOfRetries = 0) {
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏given the complexity of this function and the fact that it has its own tests suite, what about moving it to a dedicated file?

Reply:
👍